checkbox-support/checkbox_support/tests/test_timeout.py

```diff
@@ -137,38 +139,84 @@ def test_unpicklable_return_raises(self):
         def k():
             return lambda x: ...
 
-        with self.assertRaises(AttributeError):
+        with self.assertRaises(SystemExit):
             k()
 
-    @patch("checkbox_support.helpers.timeout.Process")
-    @patch("os.setsid")
-    def test_unpicklable_raise_raises(self, os_setid, process_type_mock):
+    def test_unpicklable_raise_raises(self):
         """
         The reason why this raises is that the timeout decorator pushes the
         function to another process. Trying to raise an un-picklable object
         will raise a pickle error.
         """
 
-        # this mocks process because else the coverage doesn't get the
-        # coverage
-        def init(*args, **kwargs):
-            process_type_mock.target = kwargs["target"]
-            process_type_mock.args = kwargs["args"]
-            process_type_mock.kwargs = kwargs["kwargs"]
-            return process_type_mock
-
-        def start():
-            return process_type_mock.target(
-                *process_type_mock.args, **process_type_mock.kwargs
-            )
-
-        process_type_mock.side_effect = init
-        process_type_mock.start = start
-        process_type_mock.is_alive.return_value = False
-
         @timeout(1)
         def k():
             raise ValueError(lambda x: ...)
 
         with self.assertRaises(SystemExit):
             k()
+
+    def test_timeout_kills_subprocess_tree_on_timeout(self):
+        """
+        This tests that the timeout decorator not only kills the direct child
+        (function under test) but also any sub-process it has spawned. This
+        is done because one could `subprocess.run` a long-lived process from
+        the function and it could cause mayhem (and block the test session as
+        Checkbox waits for all childs to be done)
+        """
+
+        def inner(pid_pipe):
+            pid_pipe.send(os.getpid())
+            pid_pipe.close()
+            time.sleep(1e4)
```

Comment:
This line gave me goose bumps. What happens the day the timeout decorator is broken and this line is executed while running unit tests? It might block CI runs, no?

Reply:
if the test is supposed to fail, it will fail. I'm pretty sure it wont hang the session, it could leak an inner process on failure, but there is no workaround to that, the test is to test that it will actually kill the whole tree.